### PR TITLE
Upload neighbour responses

### PR DESCRIPTION
**Pull request by cdccollins**
Add a form to bops-applicants so neighbours can upload there responses directly to BoPS

https://trello.com/c/QCtuDUfm/1875-see-the-comments-received-from-neighbours-via-a-form

screencapture-southwark-southwark-localhost-3001-planning-applications-2-neighbour-responses-2023-09-01-15_53_20
screencapture-southwark-southwark-localhost-3001-planning-applications-2-neighbour-responses-2023-09-01-15_53_36
screencapture-southwark-southwark-localhost-3001-planning-applications-2-neighbour-responses-2023-09-01-15_53_53
screencapture-southwark-southwark-localhost-3001-planning-applications-2-neighbour-responses-new-2023-09-01-15_53_09
screencapture-southwark-southwark-localhost-3001-planning-applications-2-neighbour-responses-start-2023-09-01-15_52_57
screencapture-southwark-southwark-localhost-3001-planning-applications-2-neighbour-responses-thank-you-2023-09-01-15_54_01

### CODE REVIEW

File: app/views/neighbour_responses/_about_you.html.erb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<div class="govuk-form-group">
```
Comment (benjamineskola) [inline app/views/neighbour_responses/_about_you.html.erb:1]:
```suggestion
<div class="govuk-form-group" data-controller="address-fill">
```

File: app/services/apis/os_places/client.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      def call(query)
+        faraday(query).get do |request|
+          request.options[:timeout] = TIMEOUT
+        end
+      end
+
+      private
+
+      def faraday(query)
+        @faraday ||= Faraday.new(url: "https://api.os.uk/search/places/v1/find?maxresults=20&query=#{query}&key=#{Rails.configuration.os_vector_tiles_api_key}")
```
Comment (pixeltrix) [inline app/services/apis/os_places/client.rb:19]: How is `query` validated here? Is it just the raw address info from the user?

File: app/models/neighbour_response.rb [hunk 3]
```diff
@@ -0,0 +1,124 @@
+# frozen_string_literal: true
+
+class NeighbourResponse
+  include ActiveModel::Model
+
+  STAGES = %w[about_you thoughts response check].freeze
+
+  TAGS = %i[design use privacy light access noise traffic other].freeze
+  RESPONSE_PARAMS  = [:name, :email, :address, :response, :summary_tag, TAGS, :tags, { tags: [] }].freeze
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:9]: Is the `TAGS` meant to be in the list? The resultant params looks a bit weird.

File: app/models/neighbour_response.rb [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+  PERMITTED_PARAMS = [:stage, :move_next, :move_back, :planning_application_id,
+                      { neighbour_response: RESPONSE_PARAMS }].freeze
+
+  attr_reader :params, :errors
+
+  def initialize(params)
+    @params = params.permit(*PERMITTED_PARAMS)
+    @errors = ActiveModel::Errors.new(self)
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:17]: Is there a reason why we're not using the standard Active Model setup? Including `ActiveModel::Model` sets up errors and has helpers for assigning attributes from a hash. If you include the `ActiveModel::Attributes` module you can use the `attribute` DSL to define the attributes and then use standard validation patterns, e.g.

``` ruby
class NeighbourResponse
  include ActiveModel::API
  include ActiveModel::Attributes

  attribute :name, :string
  attribute :email, :string

  validates :name, presence: true, if: -> { stage == "about_you" }
end
```

I'd also move the permitted params stuff to the more conventional controller location.

File: app/models/neighbour_response.rb [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    response_data = ActionController::Parameters.new(
+      name:, email:, address:, summary_tag:, use:,
+      privacy:, light:, access:, noise:, traffic:,
+      other:, tags:
+    )
+
+    Bops::NeighbourResponse.create(
+      params[:planning_application_id],
+      data: response_data
+    )
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:66]: No error checking?

File: app/models/neighbour_response.rb [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  end
+
+  def valid?
+    errors.clear
+    validate_response
+    errors.empty?
+  end
+
+  def validate_response # rubocop:disable Metrics/CyclomaticComplexity
+    errors.add(:name, :blank, message: "Enter your name") if stage == "about_you" && name.blank?
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:100]: We should use standard validation helpers where possible

File: app/models/neighbour_response.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    if stage == "thoughts" && summary_tag.blank?
+      errors.add(:summary_tag, :blank,
+                 message: "Select how you feel about the application")
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:104]: Messages should be specified in the locale file

File: app/models/neighbour_response.rb [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    @params = params.permit(*PERMITTED_PARAMS)
+    @errors = ActiveModel::Errors.new(self)
+  end
+
+  def stage
+    @stage ||= stage_param.in?(STAGES) ? stage_param : STAGES.first
+  end
+
+  def name
+    response_params[:name].to_s.strip
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:25]: If we use the standard Active Model patterns we can change these methods to use setters, e.g.

``` ruby
def name=(value)
  super(value.to_s.strip)
end
```

File: .rubocop.yml [hunk 9]
```diff
@@ -24,9 +24,24 @@ Metrics/BlockLength:
   Exclude:
     - spec/**/*
     - config/routes.rb
+    - app/models/neighbour_response.rb
 
 Metrics/MethodLength:
-  Max: 20
+  Max: 25
+
+Metric/AbcSize:
```
Comment (benjamineskola) [inline .rubocop.yml:32]: I wonder if we should do what we do in BOPS itself and just disable the metrics cops:

```yaml
Metrics:
  Enabled: false
```

File: app/models/neighbour_response.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
+# frozen_string_literal: true
+
+class NeighbourResponse
+  include ActiveModel::API
+  include ActiveModel::Attributes
+  include ActiveModel::Model
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:6]: ActiveModel::API and ActiveModel::Model are the same but the latter may gain features in the future so we use the former to ensure forward compatibility.

File: app/models/neighbour_response.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+  end
+
+  # rubocop:disable Metrics/PerceivedComplexity
+  # rubocop:disable Metrics/CyclomaticComplexity
+  def save
+    @stage = previous_stage and return false if moving_backwards?
+
+    @stage = stage and return false unless valid?
+
+    @stage = next_stage and return false unless done?
```
Comment (pixeltrix) [inline app/models/neighbour_response.rb:104]: Is this a clearer expression of the intended logic?

``` ruby
if moving_backwards?
  @stage = previous_stage
elsif invalid?
  @stage = stage
else
  @stage = next_stage
end

return false unless done?
```

Depends on `done?` i guess? Might need another stage added which means on the check page and the form is valid.

**Review by pixeltrix — APPROVED** [review]

**Pull request merged**